Build failure: zig

### Steps To Reproduce

Steps to reproduce the behavior:
1. run `nix-shell -p zig`

### Build log

```
...
[100%] Building stage3
dyld: lazy symbol binding faileddyld: lazy symbol binding failed: Symbol not found: ___ulock_waidyld: lazy symbol binding failedt2
  Referenced from: /tmp/nix-b: Symbol not found: ___ulock_wai: Symbol not found: ___ulock_waiuild-zig-0.11.0.drv-0/source/buit2
  Referenced from: /tmp/nix-bt2
  Referenced from: /tmp/nix-bld/zig2
  Expected in: /usr/lib/uild-zig-0.11.0.drv-0/source/buiuild-zig-0.11.0.drv-0/source/builibSystem.B.dylib

ld/zig2
  Expected in: /usr/lib/ld/zig2
  Expected in: /usr/lib/libSystem.B.dylib

libSystem.B.dylib

dyld: Symbol not found: ___ulock_wait2
  Referenced from: /tmp/ndyld: Symbol not found: ___ulockix-build-zig-0.11.0.drv-0/source_wait2
  Referenced from: /tmp/ndyld: Symbol not found: ___ulock/build/zig2
  Expected in: /usr/ix-build-zig-0.11.0.drv-0/source_wait2
  Referenced from: /tmp/nlib/libSystem.B.dylib

/build/zig2
  Expected in: /usr/ix-build-zig-0.11.0.drv-0/sourcelib/libSystem.B.dylib

/build/zig2
  Expected in: /usr/lib/libSystem.B.dylib

make[2]: *** [CMakeFiles/stage3.dir/build.make:73: stage3/bin/zig] Abort trap: 6
make[1]: *** [CMakeFiles/Makefile2:196: CMakeFiles/stage3.dir/all] Error 2
make: *** [Makefile:136: all] Error 2```
```

full logs at https://gist.github.com/CrepeGoat/d4b5f983afd3268fafe175834bfb0bf0

### Additional context

Add any other context about the problem here.

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

@andrewrk @AndersonTorres @figsoda

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-darwin"`
 - host os: `Darwin 17.7.0, macOS 10.13.6`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.20.1`
 - channels(work): `"darwin"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

-> this is a fairly old system. is it not supported? (my impression was that zig should be portable to ~all targets, as long as the target wasn't from the 1960's or something.)

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

I get a different build error on current master, though:

```
44 warnings generated.
[ 94%] Linking CXX executable zig2
[ 94%] Built target zig2
[100%] Building stage3
error: unable to create compilation: DarwinSdkNotFound
make[2]: *** [CMakeFiles/stage3.dir/build.make:73: stage3/bin/zig] Error 1
make[1]: *** [CMakeFiles/Makefile2:196: CMakeFiles/stage3.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

Hmm, I don't know why that could be. Indeed, hydra built successfully, and `nix build` downloads from the cache, but trying to reproduce the result fails with the exact same error as before.

```sh
$ checkout 92d295f
$ nix build --print-out-paths --no-link .#zig
/nix/store/divv53p9wjrsqvi3g6v3c0vabvhcb3z3-zig-0.11.0
$ nix build --rebuild .#zig
error: builder for '/nix/store/93wiqvqj0xm7bw8difdzdixjxb0rx26y-zig-0.11.0.drv' failed with exit code 2;
       last 10 log lines:
       > static inline bool zig_subo_u64(uint64_t *res, uint64_t lhs, uint64_t rhs, uint8_t bits) {
       >                                           ^
       > 44 warnings generated.
       > [ 94%] Linking CXX executable zig2
       > [ 94%] Built target zig2
       > [100%] Building stage3
       > error: unable to create compilation: DarwinSdkNotFound
       > make[2]: *** [CMakeFiles/stage3.dir/build.make:73: stage3/bin/zig] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:196: CMakeFiles/stage3.dir/all] Error 2
       > make: *** [Makefile:136: all] Error 2
       For full logs, run 'nix log /nix/store/93wiqvqj0xm7bw8difdzdixjxb0rx26y-zig-0.11.0.drv'.
$ nix-shell -p nix-info --run "nix-info -m" | grep sandbox
 - sandbox: `yes`
 ```
 
I also tried to check out the commit of [the very first successful build](https://hydra.nixos.org/build/256019393) that all the subsequent ones were cached from (3118b08635f5bd23d4daa1465571030683f38d1a), but that had the exact same result.
 
So it seems the build is very much not reproducible, meaning any change to zig's inputs might cause a rebuild that fails.

I also don't quite understand why the two store paths are different. It seems to me like they should be the same, as I'm building the same package with the same inputs, and it looks like this derivation is input-addressed.

There's some impurity here, but I can't say what it might be.

--- a/doc/hooks/zig.section.md
+++ b/doc/hooks/zig.section.md
@@ -9,14 +9,14 @@ In Nixpkgs, `zig.hook` overrides the default build, check and install phases.
 ```nix
 { lib
 , stdenv
-, zig_0_11
+, zig
 }:
 
 stdenv.mkDerivation {
   # . . .
 
   nativeBuildInputs = [
-    zig_0_11.hook
+    zig.hook
   ];
 
   zigBuildFlags = [ "-Dman-pages=true" ];

--- a/pkgs/tools/misc/clipbuzz/default.nix
+++ b/pkgs/tools/misc/clipbuzz/default.nix
@@ -3,7 +3,7 @@
 , fetchzip
 , libX11
 , libXfixes
-, zig_0_11
+, zig
 }:
 
 stdenv.mkDerivation (finalAttrs: {
@@ -15,14 +15,14 @@ stdenv.mkDerivation (finalAttrs: {
     hash = "sha256-2//IwthAjGyVSZaXjgpM1pUJGYWZVkrJ6JyrVbzOtr8=";
   };
 
-  nativeBuildInputs = [ zig_0_11.hook ];
+  nativeBuildInputs = [ zig.hook ];
 
   buildInputs = [
     libX11
     libXfixes
   ];
 
-  meta =  {
+  meta = {
     description = "Buzz on new X11 clipboard events";
     homepage = "https://trong.loang.net/~cnx/clipbuzz";
     license = lib.licenses.unlicense;

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -14765,7 +14765,9 @@ with pkgs;
 
   clipboard-jh = callPackage ../tools/misc/clipboard-jh { };
 
-  clipbuzz = callPackage ../tools/misc/clipbuzz { };
+  clipbuzz = callPackage ../tools/misc/clipbuzz {
+    zig = zig_0_12;
+  };
 
   xclip = callPackage ../tools/misc/xclip { };
 
@@ -16764,6 +16766,7 @@ with pkgs;
   };
   cargo-insta = callPackage ../development/tools/rust/cargo-insta { };
   cargo-lambda = callPackage ../development/tools/rust/cargo-lambda {
+    zig = zig_0_12;
     inherit (darwin.apple_sdk.frameworks) CoreServices Security;
   };
   cargo-limit = callPackage ../development/tools/rust/cargo-limit { };
@@ -16836,7 +16839,9 @@ with pkgs;
 
   cargo-public-api = callPackage ../development/tools/rust/cargo-public-api { };
 
-  cargo-zigbuild = callPackage ../development/tools/rust/cargo-zigbuild { };
+  cargo-zigbuild = callPackage ../development/tools/rust/cargo-zigbuild {
+    zig = zig_0_12;
+  };
 
   cauwugo = callPackage ../development/tools/rust/cauwugo { };
 
@@ -20209,6 +20214,10 @@ with pkgs;
   azmq = callPackage ../development/libraries/azmq { };
 
   babl = callPackage ../development/libraries/babl { };
+
+  backlight-auto = callPackage ../by-name/ba/backlight-auto/package.nix {
+    zig = zig_0_11;
+  };
 
   backward-cpp = callPackage ../development/libraries/backward-cpp { };
 
@@ -25169,7 +25178,7 @@ with pkgs;
   zig_0_12 = darwin.apple_sdk_11_0.callPackage ../development/compilers/zig/0.12.nix {
     llvmPackages = llvmPackages_17;
   };
-  zig = zig_0_11;
+  zig = zig_0_12;
 
   zig-shell-completions = callPackage ../development/compilers/zig/shell-completions.nix { };
 

